File: src/main/java/field/movement/MovePath.java
package field.movement;

import field.life.FieldLife;
import field.movement.fragment.*;
import lombok.Getter;
import lombok.Setter;
import util.packet.PacketReader;
import util.packet.PacketWriter;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class MovePath implements MoveFragment {

    private final List<MoveFragment> fragments;

    public MovePath(PacketReader packet) {
        this.fragments = new ArrayList<>();
        decode(packet);
    }

    @Getter @Setter Point position;
    @Getter @Setter Point vposition;

    @Override
    public void apply(FieldLife life) {
        fragments.forEach(fragment -> fragment.apply(life));
    }

    @Override
    public void decode(PacketReader packet) {
        position = packet.readPoint();
        vposition = packet.readPoint();

        byte size = packet.readByte();

        for (int i = 0; i < size; i++) {
            byte movePathAttribute = packet.readByte();

            switch (movePathAttribute) {
                case MovePathAttribute.NORMAL:
                case MovePathAttribute.HANG_ON_BACK:
                case MovePathAttribute.FALL_DOWN:
                case MovePathAttribute.WINGS:
                case MovePathAttribute.MOB_ATTACK_RUSH:
                case MovePathAttribute.MOB_ATTACK_RUSH_STOP:
                    fragments.add(new NormalMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.JUMP:
                case MovePathAttribute.IMPACT:
                case MovePathAttribute.START_WINGS:
                case MovePathAttribute.MOB_TOSS:
                case MovePathAttribute.DASH_SLIDE:
                case MovePathAttribute.MOB_LADDER:
                case MovePathAttribute.MOB_RIGHT_ANGLE:
                case MovePathAttribute.MOB_STOP_NODE_START:
                case MovePathAttribute.MOB_BEFORE_NODE:
                    fragments.add(new JumpMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.FLASH_JUMP:
                case MovePathAttribute.ROCKET_BOOSTER:
                case MovePathAttribute.BACK_STEP_SHOT:
                case MovePathAttribute.MOB_POWER_KNOCK_BACK:
                case MovePathAttribute.VERTICAL_JUMP:
                case MovePathAttribute.CUSTOM_IMPACT:
                case MovePathAttribute.COMBAT_STEP:
                case MovePathAttribute.HIT:
                case MovePathAttribute.TIME_BOMB_ATTACK:
                case MovePathAttribute.SNOWBALL_TOUCH:
                case MovePathAttribute.BUFF_ZONE_EFFECT:
                    fragments.add(new ActionMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.IMMEDIATE:
                case MovePathAttribute.TELEPORT:
                case MovePathAttribute.ASSAULTER:
                case MovePathAttribute.ASSASSINATION:
                case MovePathAttribute.RUSH:
                case MovePathAttribute.SIT_DOWN:
                    fragments.add(new TeleportMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.START_FALL_DOWN:
                    fragments.add(new StartFallDownMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.FLYING_BLOCK:
                    fragments.add(new FlyingBlockMoveFragment(movePathAttribute, packet));
                    break;
                case MovePathAttribute.STAT_CHANGE:
                    fragments.add(new StatChangeMoveFragment(movePathAttribute, packet));
                    break;
            }
        }
    }

    @Override
    public void encode(PacketWriter packet) {
        packet.writePosition(position);
        packet.writePosition(vposition);

        packet.write(fragments.size());
        fragments.forEach(fragment -> fragment.encode(packet));
    }
}

File: src/main/java/field/movement/MovePath.kt
package field.movement

import field.movement.fragment.*
import field.obj.life.FieldLife
import util.packet.PacketReader
import util.packet.PacketWriter
import java.awt.Point
import java.util.*

class MovePath(packet: PacketReader) : MoveFragment {

    private val fragments: MutableList<MoveFragment> = ArrayList()
    lateinit var position: Point
    lateinit var vposition: Point

    init {
        decode(packet)
    }

    override fun decode(packet: PacketReader) {
        position = packet.readPoint()
        vposition = packet.readPoint()

        val size = packet.readByte()
        for (i in 0 until size) {
            when (val movePathAttribute = packet.readByte()) {
                MovePathAttribute.NORMAL,
                MovePathAttribute.HANG_ON_BACK,
                MovePathAttribute.FALL_DOWN,
                MovePathAttribute.WINGS,
                MovePathAttribute.MOB_ATTACK_RUSH,
                MovePathAttribute.MOB_ATTACK_RUSH_STOP,
                -> fragments.add(NormalMoveFragment(movePathAttribute, packet))
                MovePathAttribute.JUMP,
                MovePathAttribute.IMPACT,
                MovePathAttribute.START_WINGS,
                MovePathAttribute.MOB_TOSS,
                MovePathAttribute.DASH_SLIDE,
                MovePathAttribute.MOB_LADDER,
                MovePathAttribute.MOB_RIGHT_ANGLE,
                MovePathAttribute.MOB_STOP_NODE_START,
                MovePathAttribute.MOB_BEFORE_NODE,
                -> fragments.add(JumpMoveFragment(movePathAttribute, packet))
                MovePathAttribute.FLASH_JUMP,
                MovePathAttribute.ROCKET_BOOSTER,
                MovePathAttribute.BACK_STEP_SHOT,
                MovePathAttribute.MOB_POWER_KNOCK_BACK,
                MovePathAttribute.VERTICAL_JUMP,
                MovePathAttribute.CUSTOM_IMPACT,
                MovePathAttribute.COMBAT_STEP,
                MovePathAttribute.HIT,
                MovePathAttribute.TIME_BOMB_ATTACK,
                MovePathAttribute.SNOWBALL_TOUCH,
                MovePathAttribute.BUFF_ZONE_EFFECT,
                -> fragments.add(ActionMoveFragment(movePathAttribute, packet))
                MovePathAttribute.IMMEDIATE,
                MovePathAttribute.TELEPORT,
                MovePathAttribute.ASSAULTER,
                MovePathAttribute.ASSASSINATION,
                MovePathAttribute.RUSH,
                MovePathAttribute.SIT_DOWN,
                -> fragments.add(TeleportMoveFragment(movePathAttribute, packet))
                MovePathAttribute.START_FALL_DOWN -> fragments.add(StartFallDownMoveFragment(movePathAttribute, packet))
                MovePathAttribute.FLYING_BLOCK -> fragments.add(FlyingBlockMoveFragment(movePathAttribute, packet))
                MovePathAttribute.STAT_CHANGE -> fragments.add(StatChangeMoveFragment(movePathAttribute, packet))
            }
        }
    }

    override fun apply(life: FieldLife) {
        fragments.forEach { it.apply(life) }
    }

    override fun encode(packet: PacketWriter) {
        packet.writePosition(position)
        packet.writePosition(vposition)
        packet.write(fragments.size)
        fragments.forEach { it.encode(packet) }
    }
}